File: flutter_obdii/lib/viewmodels/settings_viewmodel.dart
// Port of SettingsViewModel.swift — Jim Mittler
// ViewModel for settings screen.
// Manages connection state display, WiFi config with 500ms debounce,
// unit switching, and the connect/disconnect button action.

import 'dart:async';

import 'package:flutter/services.dart';
import 'package:package_info_plus/package_info_plus.dart';

import '../core/config_data.dart';
import '../core/obd_connection_manager.dart';
import 'base_view_model.dart';

class SettingsViewModel extends BaseViewModel {
  final SettingsConfigProviding _config;
  final OBDConnectionControlling _connection;

  // Mirrored, writable properties (bound to UI text fields / toggles)
  String wifiHost;
  int wifiPort;
  bool autoConnectToOBD;
  ConnectionType connectionType;
  MeasurementUnit units;

  OBDConnectionState connectionState;

  // Debounce timers for WiFi host/port (mirrors Swift Combine debounce .seconds(0.5))
  Timer? _hostDebounce;
  Timer? _portDebounce;

  StreamSubscription? _connectionStateSub;
  StreamSubscription? _unitsStreamSub;
  StreamSubscription? _connTypeStreamSub;

  SettingsViewModel({
    SettingsConfigProviding? config,
    OBDConnectionControlling? connection,
  })  : _config = config ?? ConfigData.instance,
        _connection = connection ?? OBDConnectionManager.instance,
        wifiHost = (config ?? ConfigData.instance).wifiHost,
        wifiPort = (config ?? ConfigData.instance).wifiPort,
        autoConnectToOBD = (config ?? ConfigData.instance).autoConnectToOBD,
        connectionType = (config ?? ConfigData.instance).connectionType,
        units = (config ?? ConfigData.instance).units,
        connectionState = (connection ?? OBDConnectionManager.instance).connectionState {
    _bindConnectionState();
    _bindExternalPublishers();
  }

  // ── Binding ─────────────────────────────────────

  void _bindConnectionState() {
    _connectionStateSub =
        _connection.connectionStateStream.listen((newState) {
      connectionState = newState;
      notifyListeners();
    });
  }

  void _bindExternalPublishers() {
    // If another part of the app changes units (e.g. CarPlay), keep in sync
    _unitsStreamSub = _config.unitsStream.listen((newUnits) {
      units = newUnits;
      notifyListeners();
    });

    _connTypeStreamSub = _config.connectionTypeStream.listen((newType) {
      connectionType = newType;
      notifyListeners();
    });
  }

  // ── WiFi Host (debounced) ────────────────────────

  void onWifiHostChanged(String newHost) {
    wifiHost = newHost;
    notifyListeners();
    _hostDebounce?.cancel();
    _hostDebounce = Timer(const Duration(milliseconds: 500), () {
      _config.wifiHost = wifiHost;
      if (connectionType == ConnectionType.wifi) {
        _connection.updateConnectionDetails();
      }
    });
  }

  // ── WiFi Port (debounced) ────────────────────────

  void onWifiPortChanged(int newPort) {
    wifiPort = newPort;
    notifyListeners();
    _portDebounce?.cancel();
    _portDebounce = Timer(const Duration(milliseconds: 500), () {
      _config.wifiPort = wifiPort;
      if (connectionType == ConnectionType.wifi) {
        _connection.updateConnectionDetails();
      }
    });
  }

  // ── Connection type ──────────────────────────────

  void onConnectionTypeChanged(ConnectionType newType) {
    if (newType == connectionType) return;
    connectionType = newType;
    _config.connectionType = newType;
    _connection.updateConnectionDetails();
    notifyListeners();
  }

  // ── Units ────────────────────────────────────────

  void onUnitsChanged(MeasurementUnit newUnits) {
    if (newUnits == units) return;
    units = newUnits;
    _config.setUnits(newUnits);
    notifyListeners();
  }

  // ── Auto connect ─────────────────────────────────

  void onAutoConnectChanged(bool value) {
    autoConnectToOBD = value;
    _config.autoConnectToOBD = value;
    notifyListeners();
  }

  // ── Connect button ───────────────────────────────

  bool get isConnectButtonDisabled =>
      connectionState == OBDConnectionState.connecting;

  void handleConnectionButtonTap() {
    switch (connectionState) {
      case OBDConnectionState.disconnected:
      case OBDConnectionState.failed:
        _connection.connect();
        break;
      case OBDConnectionState.connected:
        _connection.disconnect();
        break;
      case OBDConnectionState.connecting:
        // Do nothing while connecting
        break;
    }
  }

  // ── Diagnostics ──────────────────────────────────

  Future<Map<String, dynamic>> generateDiagnosticLogs() async {
    final info = await PackageInfo.fromPlatform();
    return {
      'timestamp': DateTime.now().toIso8601String(),
      'appVersion': '${info.version}+${info.buildNumber}',
      'connectionType': _config.connectionType.toString(),
      'units': _config.units.toString(),
      'connectionState': _connection.connectionState.toString(),
      'wifiHost': _config.wifiHost,
      'wifiPort': _config.wifiPort,
    };
  }

  // ── Number formatter (for port text field) ───────

  static final TextInputFormatter numberFormatter =
      FilteringTextInputFormatter.digitsOnly;

  // ── Dispose ──────────────────────────────────────

  @override
  void dispose() {
    _hostDebounce?.cancel();
    _portDebounce?.cancel();
    _connectionStateSub?.cancel();
    _unitsStreamSub?.cancel();
    _connTypeStreamSub?.cancel();
    super.dispose();
  }
}

File: kotlin_obdii/app/src/main/java/com/rheosoft/obdii/viewmodels/SettingsViewModel.kt
package com.rheosoft.obdii.viewmodels

import com.rheosoft.obdii.core.ConfigData
import com.rheosoft.obdii.core.ConnectionType
import com.rheosoft.obdii.core.MeasurementUnit
import com.rheosoft.obdii.core.OBDConnectionControlling
import com.rheosoft.obdii.core.OBDConnectionManager
import com.rheosoft.obdii.core.OBDConnectionState
import com.rheosoft.obdii.core.SettingsConfigProviding
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.collectLatest
import kotlinx.coroutines.launch

data class SettingsUiState(
    val wifiHost: String,
    val wifiPort: Int,
    val autoConnectToOBD: Boolean,
    val connectionType: ConnectionType,
    val units: MeasurementUnit,
    val connectionState: OBDConnectionState,
    val appVersion: String = "",
)

class SettingsViewModel(
    private val config: SettingsConfigProviding = ConfigData,
    private val connection: OBDConnectionControlling = OBDConnectionManager,
) : BaseViewModel() {
    private val scope = CoroutineScope(Dispatchers.Default + Job())
    private var hostDebounceVersion = 0L
    private var portDebounceVersion = 0L

    var wifiHost: String = config.wifiHost
        private set
    var wifiPort: Int = config.wifiPort
        private set
    var autoConnectToOBD: Boolean = config.autoConnectToOBD
        private set
    var connectionType: ConnectionType = config.connectionType
        private set
    var units: MeasurementUnit = config.units
        private set
    var connectionState: OBDConnectionState = connection.connectionState
        private set
    private var _appVersion: String = ""
    private val _uiStateStream = MutableStateFlow(buildUiState())
    val uiStateStream: StateFlow<SettingsUiState> = _uiStateStream.asStateFlow()

    init {
        scope.launch {
            connection.connectionStateStream.collectLatest {
                connectionState = it
                emitUiState()
                notifyChanged()
            }
        }
        scope.launch {
            config.unitsStream.collectLatest {
                units = it
                emitUiState()
                notifyChanged()
            }
        }
        scope.launch {
            config.connectionTypeStream.collectLatest {
                connectionType = it
                emitUiState()
                notifyChanged()
            }
        }
    }

    fun onWifiHostChanged(newHost: String) {
        wifiHost = newHost
        emitUiState()
        notifyChanged()
        val local = ++hostDebounceVersion
        scope.launch {
            delay(500)
            if (local != hostDebounceVersion) return@launch
            config.wifiHost = wifiHost
            if (connectionType == ConnectionType.wifi) connection.updateConnectionDetails()
        }
    }

    fun onWifiPortChanged(newPort: Int) {
        wifiPort = newPort
        emitUiState()
        notifyChanged()
        val local = ++portDebounceVersion
        scope.launch {
            delay(500)
            if (local != portDebounceVersion) return@launch
            config.wifiPort = wifiPort
            if (connectionType == ConnectionType.wifi) connection.updateConnectionDetails()
        }
    }

    fun onConnectionTypeChanged(newType: ConnectionType) {
        if (newType == connectionType) return
        connectionType = newType
        config.connectionType = newType
        connection.updateConnectionDetails()
        emitUiState()
        notifyChanged()
    }

    fun onUnitsChanged(newUnits: MeasurementUnit) {
        if (newUnits == units) return
        units = newUnits
        config.setUnits(newUnits)
        emitUiState()
        notifyChanged()
    }

    fun onAutoConnectChanged(value: Boolean) {
        autoConnectToOBD = value
        config.autoConnectToOBD = value
        emitUiState()
        notifyChanged()
    }

    val isConnectButtonDisabled: Boolean
        get() = connectionState == OBDConnectionState.connecting

    fun handleConnectionButtonTap() {
        when (connectionState) {
            OBDConnectionState.disconnected, OBDConnectionState.failed -> scope.launch {
                runCatching { connection.connect() }
            }
            OBDConnectionState.connected -> connection.disconnect()
            OBDConnectionState.connecting -> Unit
        }
    }

    fun setAppVersion(version: String) {
        if (_appVersion == version) return
        _appVersion = version
        emitUiState()
    }

    fun prepareLogExport(): String {
        val uiState = _uiStateStream.value
        return """
            {
              "timestamp":"${java.time.Instant.now()}",
              "connectionType":"${uiState.connectionType}",
              "units":"${uiState.units}",
              "connectionState":"${uiState.connectionState}",
              "appVersion":"${uiState.appVersion}",
              "wifiHost":"${uiState.wifiHost}",
              "wifiPort":${uiState.wifiPort}
            }
        """.trimIndent()
    }

    private fun emitUiState() {
        _uiStateStream.value = buildUiState()
    }

    private fun buildUiState(): SettingsUiState = SettingsUiState(
        wifiHost = wifiHost,
        wifiPort = wifiPort,
        autoConnectToOBD = autoConnectToOBD,
        connectionType = connectionType,
        units = units,
        connectionState = connectionState,
        appVersion = _appVersion,
    )
}
